Reworking Rating Widget

SaumiaSinghal opened the pull request:
This PR includes the work on rewriting Rating Widget.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Previously, Rating-Widget displayed stars simply on multiple lines, when the number of stars could not be accommodated in a single line. In this PR, the user just can have maximum of two lines for the rating widget.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with Rating widget like [TestRatingWidget .xml.txt](https://github.com/getodk/collect/files/4817038/TestRatingWidget.xml.txt)

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

SaumiaSinghal committed:
create rating_widget_answer_xml

SaumiaSinghal committed:
rework RatingWidget.java

SaumiaSinghal committed:
enable view binding

SaumiaSinghal committed:
add unit tests

SaumiaSinghal committed:
add unit tests

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
make unit tests green

SaumiaSinghal committed:
fix rating bar setNumOfStars

SaumiaSinghal committed:
fix lint error

SaumiaSinghal commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,116 +14,92 @@
+                this.ratingBar = binding.ratingBar2;
+            } else {
+                binding.ratingBar3.setVisibility(VISIBLE);
+                this.ratingBar = binding.ratingBar3;
```
I have three `Ratingbar` with different styles In the XML file. All of thems visibility is turned off by default. This part of code checks which `Ratingbar` needs to be displayed depending on the number of stars and screen dimensions.

seadowg commented:
@SaumiaSinghal I'm not sure about the change in behaviour here. Making the stars smaller as the total number increases does let us fit them on one line for longer but it means that each star's touch/tap area is becoming smaller and smaller. I'm worried this will make them very hard to select.

I don't have that much anecdotal evidence around how the rating widget is used (and it doesn't look like we have any analytics around it) but my assumption would be that people would typically use between 1 and 5 stars and that we don't need to optimize the aesthetics for numbers higher than that. Maybe @grzesiek2010 / @lognaturel can confirm or deny here?

seadowg commented on collect_app/build.gradle:
```diff
@@ -70,6 +70,10 @@ def mapboxToken = secrets.getProperty('MAPBOX_ACCESS_TOKEN', '')
 android {
     compileSdkVersion(28)
 
+    viewBinding {
```
It looks like Google have switched up how this should be enabled annoyingly (found [here](https://developer.android.com/topic/libraries/view-binding#setup)). This will throw up a deprecation warning on build. Instead, we need:

```
 android {
    buildFeatures {
        viewBinding true
    }
}
```

SaumiaSinghal replied:
I did the same initially, but the Gradle sync fails and it says `No signature of method: build_.android() is applicable for argument types`

seadowg replied:
Ah it looks like this PR has the older Gradle plugin. If you update your branch (`git rebase master`) so it has the latest changes from `master` you should be able to use the new definition.

SaumiaSinghal replied:
Oh sure I will do that @seadowg

seadowg commented on config/quality.gradle:
```diff
@@ -55,6 +55,7 @@ android {
         abortOnError true
         xmlReport true
         htmlReport true
+        checkGeneratedSources true
```
What was this needed for?

SaumiaSinghal replied:
I used this as otherwise `./gradlew lintDebug` fails and it gives an error for the unused resource (`geo_widget_answer.xml`)

seadowg replied:
Ah do you mean `rating_widget_answer.xml`? I was confused initially but guessing this is because of View Binding's magic. Makes sense now!

grzesiek2010 commented:
Yeah no one complained about having that widget with multiple lines of stars so there is no need to change it. As @seadowg probably in 99% of cases 5 start are used so it wasn't a problem. On the other hand if there is one crazy user which want to have let's say 100 stars probably they wouldn't be able to use the widget at all.

seadowg commented on collect_app/src/main/res/layout/rating_widget_answer.xml:
```diff
@@ -0,0 +1,37 @@
+    android:orientation="vertical"
+    android:layout_width="wrap_content"
+    android:layout_height="match_parent"
+    android:paddingStart="@dimen/margin_small"
```
All these padding values can just be replaced with one `padding` as they are all the same!

seadowg reviewed (changes requested):
A couple of comments inline but I think we should discuss the behaviour change before reviewing most of this.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,116 +14,92 @@
         DisplayMetrics dm = getScreenDimensions((Activity) getContext());
-        int widthOfStar = starDrawable.getIntrinsicWidth();
-        return (dm.widthPixels - ASSUMED_TOTAL_MARGIN_AROUND_WIDGET) / widthOfStar;
+        return (dm.widthPixels - ASSUMED_TOTAL_MARGIN_AROUND_WIDGET) / ((int) UiUtils.convertDpToPixel(widthOfStar, getContext()));
```
I think you can  use `ScreenUtils.getScreenWidth()` here and get rid of `getScreenDimensions()` at all.

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
refactor gradle file

SaumiaSinghal commented:
I agree with @seadowg and @grzesiek2010. I just changed it because I thought the appearance of `RatingBar.INDICATOR` looks better than rating bar on multiple lines when we have 8 or 9 stars, or maybe for more stars when the user has a smaller screen. However, I feel that in case of 10 stars and with `RatingBar.SMALL` style, there might be problem or the user may not be able to select the desired number of stars with as much ease as before. So, should I revert back to the old behaviour?

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,91 @@
 
-        gridLayout = new GridLayout(context);
-        gridLayout.setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
 
```
You have two blank lines here please remove one.

grzesiek2010 commented:
> So, should I revert back to the old behaviour?

I would say yes. If during your work you notice something that could be improved but is not directly related to tour goal please file a separate issue where we can discuses it. Here your goal is to refactor the widget not to implement fixes like that one.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -17,69 +17,89 @@
 
-    protected RatingWidgetAnswerBinding getBinding() {
-        return binding;
+    protected RatingBar getRatingBar() {
```
This is just for tests right? I would change access modifier (of the field) to package-private and use that field directly without any getter method like this one.

SaumiaSinghal replied:
Thanks @grzesiek2010! That's a better approach.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -17,69 +17,89 @@
     public static final int ASSUMED_TOTAL_MARGIN_AROUND_WIDGET = 40;
 
-    private RatingWidgetAnswerBinding binding;
+    private RatingBar ratingBar;
```
Why do we need this extra field at all? Why can't you use binding.ratingBar everywhere?

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/widgets/RatingWidgetTest.java:
```diff
@@ -2,60 +2,130 @@
+    }
+
+    @Test
+    public void getAnswer_whenPromptAnswerDoesNotHaveAnswer_returnsZero() {
```
Should be returns null not zero right?

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/widgets/support/QuestionWidgetHelpers.java:
```diff
@@ -37,4 +38,18 @@ public static FormEntryPrompt promptWithReadOnly() {
                 .build();
     }
+
+    public static FormEntryPrompt promptWithRangeQuestionAndAnswer(QuestionDef questionDef, IAnswerData answer) {
```
You can use it in other widgets as well so you should''t use Range word here.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/widgets/support/QuestionWidgetHelpers.java:
```diff
@@ -37,4 +38,18 @@ public static FormEntryPrompt promptWithReadOnly() {
+                .build();
+    }
+
+    public static FormEntryPrompt promptWithReadOnlyAndRangeQuestion(QuestionDef questionDef) {
```
Same as above it's not just for range widgets.

SaumiaSinghal committed:
show rating bar in two lines

SaumiaSinghal committed:
use two rating bars

SaumiaSinghal committed:
use default package private access modifier for RatingWidgetBinding

SaumiaSinghal commented:
Thanks @grzesiek2010 for reviewing. I reverted back to the original behavior. I think having the rating bar shown in two lines, not more than that, should suffice.

SaumiaSinghal committed:
fix code style

grzesiek2010 commented:
> I think having the rating bar shown in two lines, not more than that, should suffice.

Please revert this change as well we didn't discuss it so as I said above in this pr where the goal is to improve the code (not behavior) it shouldn't exist.

SaumiaSinghal commented:
Hi @grzesiek2010! I had a doubt, if I revert this change, then I might need to initialize the `Ratingbar` dynamically, as I would not know how many ratingbars would be required to display the required number of stars. And, as part of the project 'Rewriting the Widget Framework', I want to shift as much code as possible to XML files.

grzesiek2010 commented:
> I had a doubt, if I revert this change, then I might need to initialize the Ratingbar dynamically, as I would not know how many ratingbars would be required to display the required number of stars. And, as part of the project 'Rewriting the Widget Framework', I want to shift as much code as possible to XML files.

I see only creating the layout dynamically we can achieve multi-line rating bar. So I can see two options:
1. We can keep creating this widget dynamically.
2. Or we can display it in one line like all the rating bars I can google. I don't know why we didn't do this in that way since in 99.99% of cases probably our users have 5 stars. @lognaturel what do you think was there anyone who asked for such strange rating bar with many stars?

lognaturel commented:
I think the benefits of keeping the layout in XML are worth it and as @grzesiek2010 says, it's going to be really rare to have so many stars that they can't fit on one line. As @seadowg pointed out to me, [imdb](https://www.imdb.com/) is an example of a broadly-used system that has a 10-star rating. On smaller screens that would wrap in portrait. So I think @SaumiaSinghal's approach of allowing up to 2 lines is a good middle ground. 

Ideally there'd be a message if not all stars can be displayed but I don't know how easy that is. It's a problem if there are 10 stars, the user only sees 8 and selects 8 thinking that's the max rating, for example. If displaying a message is really impractical, then I think we just need to be really explicit in the documentation.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
 public class RatingWidget extends QuestionWidget {
 
     public static final int ASSUMED_TOTAL_MARGIN_AROUND_WIDGET = 40;
+    public static final int STANDARD_WIDTH_OF_STAR = 48;
```
Both can be private.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
 
-        addAnswerView(gridLayout, WidgetViewUtils.getStandardMargin(context));
-    }
+        int maxNumberOfStars = (int) ((ScreenUtils.getScreenWidth() - ASSUMED_TOTAL_MARGIN_AROUND_WIDGET)
```
Please factor it out to a separate method, I mean the calculation because you can keep the variable.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
     @Override
     public IAnswerData getAnswer() {
-        return answer != null ? new IntegerData(answer) : null;
+        return binding.ratingBar1.getRating() == 0.0F ? null :
```
I think more readable way of using the ternarry operator in this case would be:
```
return binding.ratingBar1.getRating() == 0.0F 
                        ? null 
                        : new IntegerData((int) (binding.ratingBar1.getRating() + binding.ratingBar2.getRating()));
```
so new line for each option.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
     }
 
     @Override
     public void setOnLongClickListener(OnLongClickListener l) {
```
Please fix this one-char parameter name, even simple `listener` would be better.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
-            }
-        }
+        RangeQuestion rangeQuestion = (RangeQuestion) prompt.getQuestion();
+        int numberOfStars = rangeQuestion.getRangeEnd().intValue();
```
This also could be factored out to a new method (again the calculation, you can keep the variable - as I said below).

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -14,121 +14,95 @@
-        gridLayout.setColumnCount(columns);
-        gridLayout.setRowCount(rows);
+        binding.ratingBar1.setStepSize(1.0F);
+        binding.ratingBar2.setStepSize(1.0F);
```
You already have `android:stepSize in the layout file so why do you need this here too?

SaumiaSinghal replied:
I'll fix this. Thanks for pointing out

SaumiaSinghal committed:
code refactor

kkrawczyk123 commented:
@SaumiaSinghal I have noticed issues:
- There is a color difference between master and PR. Is that change intentional?
**PR:**

| light mode | dark mode |
| ------ | ------ | 
| ![Screenshot (Jul 24, 2020 10_54_03 AM)](https://user-images.githubusercontent.com/33342488/88375690-4f76cf00-cd9c-11ea-9cfd-04e92954b317.png) | ![Screenshot (Jul 24, 2020 10_55_13 AM)](https://user-images.githubusercontent.com/33342488/88375714-5f8eae80-cd9c-11ea-89c0-179c7f6a46a7.png) |

**Master:**

| light mode | dark mode |
| ------ | ------ | 
| ![Screenshot_20200724-105833](https://user-images.githubusercontent.com/33342488/88375909-b4cac000-cd9c-11ea-98ad-e1526bd83489.png) | ![Screenshot_20200724-105740](https://user-images.githubusercontent.com/33342488/88375923-bbf1ce00-cd9c-11ea-81c6-51e4fe61c620.png) |

- The color difference on Android 5.1: Stars are gray and they have black/white frames.

| light mode | dark mode |
| -------- | -------- | 
| ![Screenshot (09_45AM, Jul 24, 2020)](https://user-images.githubusercontent.com/33342488/88371825-14bd6880-cd95-11ea-905d-890ae6aea022.png) | ![Screenshot (09_47AM, Jul 24, 2020)](https://user-images.githubusercontent.com/33342488/88371858-2272ee00-cd95-11ea-862f-8d4b1f8876a3.png) |

- I am able to choose a half or other parts of stars, see the gif:
![ezgif com-video-to-gif (14)](https://user-images.githubusercontent.com/33342488/88372596-85b15000-cd96-11ea-9052-13ada67d3701.gif)
The second issue that is visible at the gif is that answers are removed when moving forward-backward and when jumping to hierarchy. The form that I have used: [TestRatingWidget.xml.txt](https://github.com/getodk/collect/files/4970924/TestRatingWidget.xml.txt)



@getodk-bot unlabel "needs testing"

SaumiaSinghal committed:
fix stars highlight issue

SaumiaSinghal committed:
add unit test

SaumiaSinghal commented:
@kkrawczyk123 I have fixed all the mentioned issues. Sorry for multiple testing requests. Can you test again, please?

kkrawczyk123 commented:
@getodk-bot label "needs testing"

kkrawczyk123 commented:
@lognaturel I have a question about the behavior on Android 5.1 where not-chosen stars are not fulfilled and have a highlight, it differs from other devices where all not selected stars are fulfilled. Should we fix it now, open separate issue, or maybe changing it is not worth the effort as it's only Android 5 that is used rather rarely? 
See the screenshots:
Android 5.1:
| light mode | dark mode |
| ------ | ------- |
| ![Screenshot (12_52PM, Jul 27, 2020)](https://user-images.githubusercontent.com/33342488/88534809-23ac5100-d009-11ea-84ec-a6a8669ab109.png) | ![Screenshot (12_54PM, Jul 27, 2020)](https://user-images.githubusercontent.com/33342488/88534840-2ad35f00-d009-11ea-84ab-1adc725ebd98.png) |

other Androids:
| light mode | dark mode |
| ------ | ------- |
| ![Screenshot_20200724-105833](https://user-images.githubusercontent.com/33342488/88534918-51919580-d009-11ea-9c8d-8a53e94c1385.png) | ![Screenshot_20200724-105740](https://user-images.githubusercontent.com/33342488/88534933-58b8a380-d009-11ea-906a-23830acb91db.png) |

mmarciniak90 commented:
@SaumiaSinghal I noticed unexpected behaviors.

Use case 1:
1. I used this form [TestRatingWidget.xml.txt](https://github.com/getodk/collect/files/4988308/TestRatingWidget.xml.txt)
2. I opened the last question with 10 starts
3. I selected all stars.
4. Then I clicked on the fourth star -> I expected that only 4 stars will be selected and 6 will be unselected
5. Only stars in the second line are unselected

![3937](https://user-images.githubusercontent.com/3907554/88666086-dfd84b00-d0df-11ea-83bd-4a8cd7c7451f.gif)

Use case 2:
1. I used this form [TestRatingWidget.xml.txt](https://github.com/getodk/collect/files/4988308/TestRatingWidget.xml.txt)
2. I opened the last question with 10 starts
3. I selected all stars.
4. I opened remove response dialog
5. I accepted removing response
6. Only stars in the second line are unselected

@getodk-bot unlabel "needs testing"

SaumiaSinghal committed:
fix set rating problem

SaumiaSinghal commented:
Hi @mmarciniak90, Thanks for reporting the issue. I have fixed the problem. Can you test again please?

mmarciniak90 commented:
@getodk-bot label "needs testing"

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -61,8 +61,8 @@ protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int a
         binding.ratingBar1.setOnRatingBarChangeListener((ratingBar, rating, fromUser) -> {
-            binding.ratingBar1.setRating(rating);
             binding.ratingBar2.setRating(0.0F);
+            binding.ratingBar1.setRating(rating);
```
Could this change have been driven out with a test? Maybe there was one and I missed it!

SaumiaSinghal replied:
Sorry @seadowg for missing the tests. I added unit tests and confirmed that unit test`changingRating_updatesAnswer_forRatingBarInMultipleLines()` fails for the last commit, and passes on the latest one.

SaumiaSinghal committed:
add unit tests

lognaturel commented:
> Android 5.1 where not-chosen stars are not fulfilled and have a highlight, it differs from other devices where all not selected stars are fulfilled

I think this is fine to leave as-is. It's still very clear what's going on and we have fewer and fewer Android 5.1 users.

mmarciniak90 commented:
@SaumiaSinghal I noticed one more bug. It is visible only on Android 7.0. I checked the real device and emulator.

Sometimes if I select X stars, X+1 stars are selected. I hope you will be able to see this on the attached gif and reproduce it.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/3907554/88817831-492e8b80-d1be-11ea-8252-15e69e593fd0.gif)

@lognaturel I know that this problem is visible only on one Android (among those that we have) but I feel that we should fix it because it's not possible to easily select the answer you want.

mmarciniak90 commented:
@SaumiaSinghal were you able to reproduce this issue? Do you need some help?

@getodk-bot unlabel "needs testing"

SaumiaSinghal commented:
Oh sorry @mmarciniak90 Somehow I missed your comment. I will check rn

SaumiaSinghal committed:
update unt tests

SaumiaSinghal committed:
fix rating bar showing wrong rating value on Android 7

SaumiaSinghal commented:
Thanks @mmarciniak90 for reporting the issue. It has something to do with `mTouchProgressOffset` value not set correctly in `AbsSeekBar` class. I found the fix [here](https://stackoverflow.com/questions/44342481/android-nougat-7-0-api-24-rating-count-is-wrong-in-onratingbarchangelistener). Can you test again please?

mmarciniak90 commented:
@getodk-bot label "needs testing"

mmarciniak90 commented:
@SaumiaSinghal there is still unexpected behavior:

1. Open _TestRatingWidget.xml_ form
2. Go to question with 10 stars
3. Select all stars
4. Click on the last star in the first row to unselect stars in the second row -> it's not possible

I checked this case on Android versions: 7.0, 8.1, 9.0

![07-08-2020-09-45(1)](https://user-images.githubusercontent.com/3907554/89622938-1c503780-d894-11ea-9c0b-117eed58306f.gif)

Additionally, I still see that 1 star more is marked than I selected on Android 7.

@getodk-bot unlabel "needs testing"

SaumiaSinghal committed:
add touch listener to change rating in other rating bar simulataneously

SaumiaSinghal commented:
Hi @mmarciniak90! I fixed the problem with selecting rating when clicking on the last star of the `RatingBar`. I should have guessed that, that is the case when `OnRatingChangeListener` is not called as the rating on the top rating bar is still the same. I fixed the problem by setting an `OnTouchListenr` on both the rating bars, which change the rating of the other rating bar as soon as the rating of any one of them is changed (anyone of them is touched).  
However, I didn't see the issue of setting a rating on Android 7. I do not have a real device, but I tested it on an emulator.

seadowg reviewed (changes requested):
@SaumiaSinghal Sorry I think this got a little lost. Could you rebase on top of master?

seadowg commented on collect_app/build.gradle:
```diff
@@ -70,8 +70,8 @@ def mapboxToken = secrets.getProperty('MAPBOX_ACCESS_TOKEN', '')
-    viewBinding {
-        enabled = true
+    buildFeatures {
+        viewBinding true
```
I don't think we need this anymore right? It's enabled in the `android` block now.

seadowg reviewed (approved):
@getodk-bot label "needs testing"

One little tweak but doesn't affect this going to QA!

kkrawczyk123 commented:
@SaumiaSinghal I have noticed another issue. On the rating widget where stars do not fit to one line I have selected a whole 1st line and at least 1 star from the 2nd line. After rotating the screen all-stars are selected, see the gif:
![ezgif com-video-to-gif (34)](https://user-images.githubusercontent.com/33342488/93989740-98420880-fd8a-11ea-8ba5-216f5ece4cfa.gif)

When I opened hierarchy view it says that I have selected 11 stars from 10:
![Screenshot_2020-09-23-10-45-32-219_org odk collect android](https://user-images.githubusercontent.com/33342488/93989793-a42dca80-fd8a-11ea-901f-2bcf394caf26.jpg)

@getodk-bot unlabel "needs testing"

seadowg commented:
@SaumiaSinghal looks like there is a problem that QA found here and some small conflicts. Sorry for this being so slow. Trying to go through these in order!

SaumiaSinghal committed:
set rating bar style

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
update unit tests

seadowg commented:
@SaumiaSinghal could you try rerunning on Circle CI. It won't let me do it 😭

SaumiaSinghal commented:
Sure @seadowg! Also, I could not fix the problem reported by the QA. I tried debugging the app and commenting out my code, but could not find where my code is setting the rating to max number of stars on screen rotation.

seadowg commented:
Ach I think this is due to a super annoying, almost completely invisible feature of Android. Basically when the Activity rotates and is recreated it will take the state from views with IDs, save them and then restore them once the view is rendered again. This means that the state from `rating_bar1` is being saved (100%) and then is restored when the view is rotated. This happens after the view has been created so after we've set the rating bar to have the correct number of starts. We ran into this recently with selects in field lists as well. 

The solution is to add `android:saveEnabled="false"` to the `RatingBar` views so this automatic save/restore doesn't happen. Yuck! I'll add a note to the docs about this.

SaumiaSinghal committed:
set saveEnabled to false to prevent rating bar from retaining values on screen rotation

SaumiaSinghal commented:
Thanks so much @seadowg, it worked. It is ready for testing

mmarciniak90 commented:
@SaumiaSinghal I can still see the issue described here: https://github.com/getodk/collect/pull/3937#issuecomment-665727700

It is a gif from the current version

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/3907554/97286937-48040d80-1844-11eb-875e-2af346ff64a2.gif)

@kkrawczyk123 are you able to reproduce it on Android 7.1?

SaumiaSinghal commented:
Hi @mmarciniak90 I didn't notice this bug on my emulator running Android 7.0 :/

![rating_widget](https://user-images.githubusercontent.com/35730054/97560432-a7108080-1a04-11eb-84df-d8ef192cfdb9.gif)

kkrawczyk123 commented:
@mmarciniak90 @SaumiaSinghal I am not able to reproduce it on my Android 7.1.

seadowg commented:
@mmarciniak90 what version of Android are you seeing that problem on?

mmarciniak90 commented:
Only on a real device with Android 7.0, Samsung Galaxy S7. 
I think creating a separate issue for this is a good solution. We should not block this PR because of this. 
@lognaturel @seadowg are you agree?

EDIT:
New issue created #4192

kkrawczyk123 commented:
**Tested with success!**
Verified on Androids: 5.1, 7.1, 8.1, 9.0 and 10.

Tested scenarios:
- light/dark mode
- remove response
- cancel removing response
- screen rotation 
- minimize app
- jump form hierarchy
- going backward/forward
- RTL language 
- all detected issues have been fixed 

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"
